projects/centre_console/src/race_switch.c

```diff
@@ -0,0 +1,146 @@
+static GpioSettings s_voltage_enable_gpio_settings = { .state = GPIO_STATE_LOW,
+                                                       .direction = GPIO_DIR_OUT,
+                                                       .resistor = GPIO_RES_NONE,
+                                                       .alt_function = GPIO_ALTFN_NONE };
+
+static InterruptSettings s_interrupt_settings = { .type = INTERRUPT_TYPE_INTERRUPT,
+                                                  .priority = INTERRUPT_PRIORITY_NORMAL };
+
+static void prv_voltage_monitor_error_callback(VoltageRegulatorError error, void *context) {
+  if (error == VOLTAGE_REGULATOR_ERROR_ON_WHEN_SHOULD_BE_OFF) {
+    LOG_WARN("Voltage Regulator Error: On when should be off");
+  } else {
+    LOG_WARN("Voltage Regulator Error: Off when should be on");
+  }
+}
+
+static VoltageRegulatorSettings s_voltage_regulator_settings = {
+  .enable_pin = VOLTAGE_REGULATOR_ENABLE_ADDRESS,
+  .monitor_pin = VOLTAGE_REGULATOR_MONITOR_ADDRESS,
+  .timer_callback_delay_ms = VOLTAGE_REGULATOR_DELAY_MS,
+  .error_callback = prv_voltage_monitor_error_callback,
+  .error_callback_context = NULL,
+};
+
+FSM_DECLARE_STATE(race_switch_off);
+FSM_DECLARE_STATE(race_switch_on);
+
+FSM_STATE_TRANSITION(race_switch_off) {
+  FSM_ADD_TRANSITION(RACE_STATE_ON, race_switch_on);
+}
+
+FSM_STATE_TRANSITION(race_switch_on) {
+  FSM_ADD_TRANSITION(RACE_STATE_OFF, race_switch_off);
+}
+
+static RaceState s_event_lookup[] = { [RACE_SWITCH_EVENT_OFF] = RACE_STATE_OFF,
+                                      [RACE_SWITCH_EVENT_ON] = RACE_STATE_ON };
+
+// Triggered when the fsm switches to the normal mode
+// 5V regulator is enabled
+static void prv_state_race_off_output(Fsm *fsm, const Event *e, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  voltage_regulator_set_enabled(&storage->voltage_storage, true);
+  storage->current_state = RACE_STATE_OFF;
+}
+
+// Triggered when the fsm switches to the race mode
+// 5V regulator is disabled
+static void prv_state_race_on_output(Fsm *fsm, const Event *e, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  voltage_regulator_set_enabled(&storage->voltage_storage, false);
+  storage->current_state = RACE_STATE_ON;
+}
+
+bool race_switch_fsm_process_event(RaceSwitchFsmStorage *storage, Event *e) {
+  return fsm_process_event(&storage->race_switch_fsm, e);
+}
+
+static void prv_gpio_interrupt_handler(const GpioAddress *address, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  GpioState input_state;
+  gpio_get_state(address, &input_state);
+
+  Event e = { 0 };
+
+  // Rising edge indicates the gpio state has become high and the car should go into race mode
+  if (input_state == GPIO_STATE_HIGH) {
+    e.id = s_event_lookup[RACE_SWITCH_EVENT_ON];
+  } else {
+    e.id = s_event_lookup[RACE_SWITCH_EVENT_OFF];
+  }
+
+  race_switch_fsm_process_event(storage, &e);
```

Comment:
Sorry, didn't catch this in the earlier review: What we usually do in module code is raise the event with `event_raise` (or `event_raise_no_data`); then the fsm will transition when the event is processed with `race_switch_fsm_process_event` in the main event loop. This way the transitioning etc happens out of the interrupt, and it gives other events time to process.

To fix your tests to comply with this, you can declare an `Event e`, use `event_process(&e)` after triggering the interrupt to get the event, and then call `race_switch_fsm_process_event` (even tie it up in a helper function if you want).